io_ogre/ogre/node_anim.py

```diff
@@ -0,0 +1,144 @@
+        logger.info('+ NLA track: %s' % nla.name)
+
+        for strip in nla.strips:
+            action = strip.action
+            actions[ action.name ] = [action, strip.action_frame_start, strip.action_frame_end]
+            logger.info('  - Action name: %s' % action.name)
+            logger.info('  -  Strip name: %s' % strip.name)
+
+    actionNames = sorted( actions.keys() )  # output actions in alphabetical order
+    for actionName in actionNames:
+        actionData = actions[ actionName ]
+        action = actionData[0]
+        anim.action = action  # set as the current action
+        write_animation( ob, actionName, actionData[1], actionData[2], doc, xmlnode )
+
+    # restore these to what they originally were
+    anim.action = savedAction
+    anim.use_nla = savedUseNla
+
+    logger.info('- Done at %s seconds' % util.timer_diff_str(start))
+
+# This function processess the objects animation frame by frame, insted of using its keyframes
+# The advantage of this method is that it respects the users  tuning of the node animation curves
+# The disadvantage is that it generates more data than method 1 and might clutter the .scene file which is not a binary file so it might take a lot longer to process when loading the .scene
+def write_animation(ob, action_name, frame_start, frame_end, doc, xmlnode):
+
+    _fps = float( bpy.context.scene.render.fps )
+
+    # Actually in Blender this does not make sense because there is only one possible animation per object, but lets maintain compatibility with Easy Ogre Exporter
+    aa = doc.createElement('animations')
+    xmlnode.appendChild(aa)
+    
+    a = doc.createElement('animation')
+    a.setAttribute("name", "%s" % action_name)
+    a.setAttribute("enable", "false")
+    a.setAttribute("loop", "false")
+    a.setAttribute("interpolationMode", "linear")
+    a.setAttribute("rotationInterpolationMode", "linear")
+    a.setAttribute("length", '%6f' % ((frame_end) / _fps))
+    aa.appendChild(a)
+    
+    frame_current = bpy.context.scene.frame_current
+    
+    initial_location = mathutils.Vector((0, 0, 0))
+    
+    for frame in range(int(frame_start), int(frame_end) + 1):
+        kf = doc.createElement('keyframe')
+        kf.setAttribute("time", '%6f' % (frame / _fps))
+        a.appendChild(kf)
+
+        bpy.context.scene.frame_set(frame)
+        
+        translation = mathutils.Vector((0, 0, 0))
+
+        if frame == frame_start:
+            initial_location = util.swap( ob.matrix_local.to_translation() )
+        else:
+            translation = util.swap( ob.matrix_local.to_translation() ) - initial_location
+        
+        t = doc.createElement('position')
+        t.setAttribute("x", '%6f' % translation.x)
+        t.setAttribute("y", '%6f' % translation.y)
+        t.setAttribute("z", '%6f' % translation.z)
+        kf.appendChild(t)
+        
+        q = doc.createElement('rotation')
+        rotation_quaternion = util.swap( ob.matrix_local.to_quaternion() )
```

Comment:
did you test this? according to how I read the ogre code, the initial_rotation should be removed here, similarly to intial_location above

Reply:
I suspected that was the case, because in one of the tests I'm doing now it seems that it is affecting me.

In fact I was about to get to work on this right now.

The problem is that I'm not sure how to do that with quaternions, I can understand it for positions/translations but it is harder to reason with quaternions. I usually have to consult the quaternion primer from the wiki XD.

Reply:
Solved

Reply:
> The problem is that I'm not sure how to do that with quaternions, I can understand it for positions/translations but it is harder to reason with quaternions. 

I tend to think of everything in matrix notation, e.g.:
```py
t = initial_location.inverse() @ location
r = initial_rotation.inverse() @ rotation
```
of course the above can be rewritten as
```py
t = -initial_location + location
```